Add support for PolyLoss (ICLR 2022)

acforvs opened the pull request:
Closes #466 

I focused on implementing the Poly-1 cross entropy loss with alpha-label smoothing since, as it is shown in the paper, the first term plays the most important role. Also, I crafted a small test showing that the behavior is indeed similar to the one of the `softmax_cross_entropy` when `\epsilon = 0`

I didn't add it to `optax/__init__.py` and `docs/api.rst` files yet (would it be ok to do that in a separate PR?)

Please, let me know what you think!

mkunesch commented on optax/_src/loss_test.py:
```diff
@@ -496,5 +496,85 @@ def test_batched(self):
+
+  @chex.all_variants
+  @parameterized.parameters(
+      dict(eps=2, alpha=0, expected=4.531657285679147),
```
Since we are testing to atol 1e-4 we can remove some of the figures from the expected results.

mkunesch reviewed (commented):
Thanks a lot for this contribution - this looks great! Thanks especially for adding the maths and recommended parameters.

I've added a few comments. I'll do another pass to compare the maths to the paper later.

acforvs commented on optax/_src/loss.py:
```diff
@@ -519,3 +519,58 @@ def hinge_loss(predictor_outputs: chex.Array,
+    """
+  chex.assert_type([logits], float)
+
+  smoothed_labels = smooth_labels(labels=labels, alpha=alpha)
```
I agree, this is indeed an independent functionality; I removed it and fixed the tests accordingly

acforvs commented on optax/_src/loss.py:
```diff
@@ -519,3 +519,58 @@ def hinge_loss(predictor_outputs: chex.Array,
+        - It is also recommended to adjust this value
+          based on the task and dataset at hand. For example, one can use
+          simple grid search to achieve it.
+      alpha: The smoothing factor, the greedy category with be assigned
```
My bad, I copypasted the description from [here](https://github.com/deepmind/optax/blob/master/optax/_src/loss.py#L107) without going through it 

Since I removed the smoothing itself, this issue is now fixed in this PR. Should I submit a separate PR to fix this line for the `smooth_labels` as well?

acforvs commented:
Thanks for the review, I updated the code!

mkunesch reviewed (approved):
Hi! Sorry for dropping the ball on this! I've added a few small nits, but we we can also address those during merging so I've also approved.

acforvs commented:
@mkunesch addressed the comments. The tests seem to be failing locally though, but it looks like the problem is in `ctc_loss_with_forward_probs`

I'll do another pass in a separate PR to fix the style in other docstrings if that's ok

mkunesch commented:
Hi! Thanks a lot for the changes and sorry for the delay!

> The tests seem to be failing locally though, but it looks like the problem is in ctc_loss_with_forward_probs
Yes, that was an unrelated problem. We've fixed that error now - could you sync with master? That should make the checks pass so that we can merge. 

> I'll do another pass in a separate PR to fix the style in other docstrings if that's ok
Sounds good, thanks a lot!

acforvs commented:
Hi, no worries!
I synced my local branch with master, but it didn't trigger the tests for some reason - could you re-run them please?

acforvs commented:
Hi @mkunesch, thanks for the review! Should I also update https://github.com/deepmind/optax/blob/master/docs/api.rst#L494 and https://github.com/deepmind/optax/blob/master/optax/__init__.py#L96?

mkunesch commented:
Ah, yes please - I forgot to check that in my review. Thanks a lot!

acforvs commented:
Added here: https://github.com/deepmind/optax/pull/537
